ambry-clustermap/src/main/java/com.github.ambry.clustermap/AmbryPartition.java

```diff
@@ -84,6 +84,24 @@ public boolean isEqual(String other) {
     return id.toString().equals(other);
   }
 
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+
+    AmbryPartition partition = (AmbryPartition) o;
+    return id.equals(partition.id);
+  }
+
+  @Override
+  public int hashCode() {
+    return id.hashCode();
```

Comment:
Did we find a case where a clustermap was returning a non-singleton AmbryPartition instance to require a hashcode? Not necessarily against having it since this can be a big gotcha when writing cluster aware code or test cases. However, if there are cases where PartitionId is found to have multiple instances for the same partition number, we should track them down since this class holds some state regarding sealed/stopped replicas.

Reply:
I'm not sure.  The frequency of cache misses (even repeated ones causing 3 retries to fail) made me suspicious but I have no proof of that.